### PR TITLE
fix(booking): carry `{*}` into application instead of resolving it away

### LINKED ISSUE #2068: [Bug]: `{*}` cost-merge reduction books but cannot apply — E4001 on a lot the account holds
### Description

A `{*}` (cost-merge) reduction is booked successfully and then fails to apply,
so `rledger check` reports `E4001 No matching lot` for a lot the account plainly
holds. `Inventory::reduce_merge` itself works — it is covered by
`test_reduce_merge_operator`, `test_reduce_merge_single_lot` and
`test_reduce_merge_three_lots` — but the feature is unreachable end to end.

Found while reviewing #2058; it is unrelated to that fix and reproduces
identically before it.

### Steps to Reproduce

```beancount
option "operating_currency" "USD"
2000-01-01 open Assets:Stock X "STRICT"
2000-01-01 open Assets:Cash USD
2000-01-01 open Income:PnL

2024-01-01 * "buy lot 1"
  Assets:Stock  10 X {100.00 USD}
  Assets:Cash  -1000.00 USD

2024-01-02 * "buy lot 2"
  Assets:Stock  10 X {120.00 USD}
  Assets:Cash  -1200.00 USD

2024-02-01 * "sell with the wildcard merge operator"
  Assets:Stock  -5 X {*}
  Assets:Cash   600.00 USD
  Income:PnL
```

`rledger check` → `error[BOOK]: No matching lot for X in Assets:Stock`

### Expected Behavior

The reduction books against the merged weighted-average lot, as
`Inventory::reduce_merge` already implements.

### Root cause (verified, not inferred)

Booking is two-phase and the merge is resolved in phase 1 against a working
copy, then re-matched in phase 2 against live state that was never merged.

Instrumenting `BookingEngine::book` on the file above:

```
BOOK OK; rewritten cost = CostSpec { number: Some(PerUnit { value: 110 }),
                                     currency: Some("USD"),
                                     date: None, label: None, merge: false }
APPLY FAILED: No matching lot for X in Assets:Stock
```

`book()` succeeds, computes the weighted average (110 = the merge of 100 and
120), and rewrites the posting to a concrete per-unit cost **with `merge:
false`**. `apply()` then re-runs the reduction against the real inventories,
which still hold the original lots at 100 and 120 — nothing is at 110, so the
match fails.

This is the same phase-1/phase-2 shape as #2048: phase 1 rewrites the posting
into something phase 2 cannot match. #2048 was a precision mismatch on a cost
that did exist; here the cost never exists outside the working copy, because
the merge that would create it is not carried into `apply`.

### Severity

Low. Python beancount rejects `{*}` too — `bean-check` on the same file says
`Cost merging is not supported yet` — so this is not a compatibility
divergence, and no ledger that works with beancount is affected. The cost is
that rledger accepts the syntax, implements the semantics at the inventory
level, and then fails with a misleading message rather than either working or
saying it is unsupported.

### Possible directions

1. Carry the merge into phase 2 — leave `merge: true` on the rewritten spec so
   `apply` performs the same merge, rather than looking for its result.
2. Have phase 1 hand phase 2 the matched lot identity instead of a cost to
   re-match, which is the general fix for this whole class (#2048 was the other
   instance).
3. Reject `{*}` explicitly with a clear diagnostic, matching beancount's
   position, and drop the inventory-level implementation to unused-code status.

### Version

0.21.0 (reproduced on `f3ac855ca2` and on `e3123ad7dd`)

### Operating System

Linux

### PR DESCRIPTION
**Pull request by robcohen**
Fixes #2068.

## The bug

`{*}` is the one cost spec that is an **operation** rather than a **filter**.

Every other spec — `{100.00 USD}`, `{2024-01-01}`, `{"lot-a"}` — *selects* among lots that already exist, so booking can resolve it into the single lot it picked and clear the spec. `{*}` *merges* the pool and only then reduces it.

Booking resolved `{*}` into the per-unit cost of the pool it would create and cleared the marker (`merge: false`). That names a lot which does not exist until the merge has run, so the application pass reported `No matching lot` for holdings the account plainly had:

```beancount
2024-01-01 * "buy lot 1"
  Assets:Stock  10 X {100.00 USD}
  Assets:Cash  -1000.00 USD

2024-01-02 * "buy lot 2"
  Assets:Stock  10 X {120.00 USD}
  Assets:Cash  -1200.00 USD

2024-02-01 * "sell against the merged pool"
  Assets:Stock  -5 X {*}     ; booked to {110.00 USD} — a lot nobody holds
  Assets:Cash   600.00 USD
  Income:PnL
```

`rledger check` → `No matching lot for X in Assets:Stock`. beancount accepts this ledger.

## The fix

The single-lot rewrite carries `cost_spec.merge` through, so application **re-runs the merge** rather than searching for its result. The account ends holding one merged pool at the weighted average (`15 X @ 110`), which is what `{*}` means.

`reduce_merge` returns a single synthetic matched position *by design* (its comment says so: expanding into one posting per original lot "would be incorrect for `{*}`"), so a `{*}` reduction can never reach the multi-lot expansion branch. This one site is the whole fix.

## The cost of carrying an operation, and the check that pays it

A carried operation is only meaningful against the inventory it was booked against. Applied to different state it would silently build a *different* pool — the outcome this codebase treats as worse than any error.

So booking records the per-unit cost of the pool it computed, and `apply_posting` compares that against what the re-executed merge actually produced, reporting a new `BookingError::MergeMismatch` on a disagreement. A merge is now either the pool booking saw or a loud error; it is never a quietly different pool.

## Verification

Three tests, each shown to be load-bearing by mutating the one changed line back to `merge: false`:

| test | crate | catches |
|---|---|---|
| `a_wildcard_merge_loads_without_errors` | loader | the reported bug, end to end through `load` |
| `a_wildcard_merge_books_and_applies_leaving_the_pool` | booking | the engine holds exactly `15 X @ 110` |
| `a_merge_applied_against_different_inventory_is_reported` | booking | the checksum fires when the pool moved |

Reverting *only* `merge: cost_spec.merge` fails all three; disabling *only* the checksum fails the third alone. Both halves are independently necessary.

Gates: `cargo test --workspace --all-features` 143 suites green; `cargo +1.97.0 clippy --workspace --all-features --all-targets -D warnings`, `cargo fmt --check`, and `RUSTDOCFLAGS=-D warnings cargo doc` all clean.

## Surfaces that see the flag

`rledger format` is source-faithful (unchanged), and BQL renders the resolved `Cost` from the inventory, not the `CostSpec` — so `-5 X {110.00 USD}`, no `*` leaking into user-facing output. The flag reaches only `meta.hash` and the FFI `CostSpec` passthrough, where carrying it is the more faithful representation. Any hash drift is confined to `{*}` ledgers, which today error out rather than load.

## Not in scope

Whether an explicit `{*}` should *realize* the merged pool in reports (BQL still shows the journal view `10 X {120}  10 X {100}  -5 X {110}`, which nets to the same 15 X / 1650 USD) is ADR-0007 behaving as designed — realization keys off `AVERAGE` accounts. That is the separate question raised on #2068.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_012nNGPA44Dt32NyxWem26xr

### CODE REVIEW

**Comment by robcohen** [comment]
Added one more assertion after review of my own diff.

The mismatch is detected *after* `reduce` has already merged and drained the lots, so reporting it correctly depends on `apply`'s undo log (#2067) putting the pool back. A cost-bearing negative posting always sets `has_reduction`, so the log is open — but a half-merged inventory is precisely the silent corruption this check exists to prevent, so the test now asserts all three lots survive unmerged and undrained rather than assuming it. Mutating `has_reduction` to `false` fails that assertion.

Two boundary cases worth recording, both erring toward not firing:

- `CostNumber::per_unit()` returns `None` for `Total` and `Compound`, so a `{{550.00 USD, *}}` spec that booking did not rewrite skips the check rather than comparing against an effective per-unit it cannot know yet.
- A raw, never-booked `{*}` reaching `apply` directly has `number: None` and also skips — nothing was recorded, so there is nothing to check against.

`Decimal`'s `PartialEq` scale-aligns, so a recorded `110` and a produced `110.00` compare equal and do not false-fire.

File: crates/rustledger-core/src/inventory/mod.rs [hunk 1]
```diff
@@ -211,6 +232,16 @@ impl From<OverflowError> for BookingError {
...
+            Self::MergeMismatch {
+                currency,
+                expected,
+                got,
+            } => write!(
+                f,
+                "{{*}} merge for {currency} produced a pool cost of {got}, but \
+                 booking recorded {expected}: this posting was applied against \
+                 different inventory than it was booked against"
+            ),
```
[the remark this answers is not in the document]
Comment (robcohen) [reply crates/rustledger-core/src/inventory/mod.rs:248]: Fixed in 4ce9631534. `MergeMismatch` now carries `Amount`s rather than bare `Decimal`s, so the message reads `110.00 USD` — and comparing `Amount`s means a pool that came back in the wrong cost currency is now a mismatch too, which the bare-number compare would have passed. `currency` stays the commodity being reduced, which is the context the message needs to name the account's holding.

File: crates/rustledger-booking/src/book.rs [hunk 2]
```diff
@@ -732,8 +744,33 @@ impl BookingEngine {
         if inv.is_booking_reduction(units, posting.cost.as_ref(), method) {
             // `reduce` only errors when the lot it would match is missing — a
             // "must book first" precondition violation.
-            inv.reduce(units, posting.cost.as_ref(), method)
+            let result = inv
+                .reduce(units, posting.cost.as_ref(), method)
                 .map_err(|e| convert_core_booking_error(e, &posting.account))?;
```
[the remark this answers is not in the document]
Comment (robcohen) [reply crates/rustledger-booking/src/book.rs:818]: Right, and following it turned up a false positive I had shipped.

`apply_posting` is `pub` and the query executor calls it directly to replay a **filtered** transaction stream. There a different pool is the *correct* answer for the stream it was given, not evidence of corruption — so the checksum aborted any query whose `FROM` filter excluded one of the merged lots:

```
$ rledger query merge.beancount \
    "select account, account_balance from not (narration ~ 'lot 1') where account ~ 'Stock'"
error: evaluation error: Assets:Stock: {*} merge for X produced a pool cost of 120.00,
       but booking recorded 110.00
```

Fixed in 4ce9631534 by moving the check to `apply` — the caller whose contract is "reproduce booking's decision" — and running it *before* the merge mutates, as you suggested, rather than after. `apply_posting`'s contract goes back to "`reduce` only errors when the lot it would match is missing", and the comment is accurate again.

Knowing the pool without building it needed the selection half of `reduce_merge` callable from `&self`, so `{*}` now has the plan/commit split (#2061): `plan_merge` computes which lots merge and at what average, `reduce_merge` commits it, `merged_pool_cost` exposes the read-only half. One implementation — perturbing the plan's pool fails six `reduce_merge` tests, so the commit half genuinely consumes it.

It is checked per posting rather than as a pre-pass, since an earlier posting of the same transaction can legitimately change the pool a later `{*}` sees. New test `a_from_filter_that_drops_a_merged_lot_still_answers` pins the query path.

**Comment by robcohen** [comment]
## Deep review of my own diff

Three defects, two of them panics. Each fix has a test that fails without it (61852481d2).

**1. The check fired on postings that never merge.** `apply_posting` reduces only when `is_booking_reduction` says so; an augmentation carrying an explicit `{110.00 USD, *}` is *added*. Comparing it against a pool it will not build rejected a valid posting. Now gated on that same canonical predicate.

**2. Rejecting a `{*}` that closes a short panicked.** A reduction is not the same thing as a negative posting — closing a short reduces cost-bearing lots with *positive* units. `rollback_needed` counts by sign (deliberately: reading current state to answer it is what caused the #2067 panic), so such a posting reached `apply`'s "the guard is unsound" assert with no undo log open:

```
thread 'probe' panicked at book.rs:1084:
rollback_needed() returned false but posting application failed
({*} merge of X would produce a pool cost of 100 USD, but booking recorded 110 USD)
```

`has_reduction` now counts `{*}` postings in both directions — sign- and state-independent, and only `{*}` widens the gate, so #1897's snapshot regression stays closed.

**3. It rejected a buy and a `{*}` sale in ONE transaction.** Booking books every posting against the inventory as it stood *before* the transaction; `apply` mutates sequentially. The earlier buy legitimately moves the pool the sale meets. The check now runs only for an account the transaction has not already touched — the state booking actually saw.

## One thing this PR does not fix

That third case exposes a book-vs-apply disagreement that is not merge-specific but is only *observable* through `{*}`:

```beancount
2024-02-01 * "buy and merge-sell together"
  Assets:Stock   10 X {120.00 USD}
  Assets:Stock   -5 X {*}
```

Booking records the reduction at **100.00** (the pre-transaction pool), while the engine's inventory ends at **15 X @ 110** (the sequential pool). The journal view totals 1,700 USD of basis; the engine holds 1,650. Beancount's semantics book against the pre-transaction state, which suggests booking is right and `apply` merges too much.

I have left it alone rather than adjudicate it inside a bug fix — on `main` this ledger fails to load at all, so nothing regresses. Happy to file it as its own issue with this repro if you want it tracked.

Gates: 145 suites green, clippy/fmt/doc clean.

**Comment by robcohen** [comment]
Filed the ordering divergence as #2070.

Worth noting what turned up while writing it up: it is **not** `{*}`-specific and **not** introduced by this PR. `AVERAGE` booking has the same shape and reproduces on `main` today with no `{*}` anywhere:

```beancount
2024-02-01 * "buy then sell in ONE transaction"
  Assets:Stock   10 X {120.00 USD}
  Assets:Stock   -5 X {}
```

on an `"AVERAGE"` account — booked postings aggregate to `15 X {113.33 USD}` (1,700 USD basis) while the engine inventory holds `15 X @ 110` (1,650). `rledger check` reports no errors either way.

Both mergers re-derive a pool from whatever the account holds when they run, so both see the lot an earlier posting of the same transaction added; booking saw only the pre-transaction state. Specs that name a concrete lot are unaffected.
